File: Assignment1.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_set>

using namespace std;

struct Point{
    int x;
};

struct Edge {
    int x;
    int y;
};

// Graph class
class Graph {
private:
    int V; // Number of vertices
    vector<vector<int>> adjList; // Adjacency list

public:
    // Constructor
    Graph(int vertices) : V(vertices) {
        adjList.resize(V);
    }

    // Function to add an edge to the graph
    void addEdge(int u, int v) {
        adjList[u].push_back(v);
        // For undirected graph, add this line:
        // adjList[v].push_back(u);
    }
};

int main() {
    std::cout << "Enter Number of verticies and tuples of edges \n";
    std::cout << "For Example: 5 (1,2) (3,4) (3,5) (4,5)\n";
    int verticies = 5;
    // std::cin >> verticies;

    std::vector<Edge> edges; // Vector to store edges
    std::string input;
    std::getline(std::cin, input);

    std::string currentEdge;
    for (char c : input) {
        if (isdigit(c) || c == '-') {
            currentEdge += c;
        } else if (c == ',') {
            edges.back().x = std::stoi(currentEdge);
            currentEdge = "";
        } else if (c == ')') {
            edges.back().y = std::stoi(currentEdge);
            currentEdge = "";
        } else if (c == '(') {
            edges.emplace_back();
        }
    }

    // Printing the edges
    std::cout << "You entered the following edges:" << std::endl;
    for (const auto& edge : edges) {
        std::cout << "(" << edge.x << ", " << edge.y << ")" << std::endl;
    }

    // Create a vector to store points
    std::vector<Point> points;

    // Loop to create vertices
    for (int i = 1; i < verticies + 1; ++i) {
        Point p;  // Create a new Point instance
        p.x = i;  // Assign the x coordinate (assuming it's based on the loop index)
        points.push_back(p);  // Add the point to the vector
    }

    std::cout << "You entered the following points:" << std::endl;
    for (const auto& point : points) {
        std::cout << point.x << std::endl;
    }

    // Create a graph
    Graph g(verticies); // Graph with vertices
    
    return 0;
    }

File: Assignment1.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

class Point {
    int x;
}

class Edge {
    int x;
    int y;
}

// Graph class
class Graph {
    private int V; // Number of vertices
    private ArrayList<ArrayList<Integer>> adjList; // Adjacency list

    // Constructor
    Graph(int vertices) {
        V = vertices;
        adjList = new ArrayList<>(V);
        for (int i = 0; i < V; ++i)
            adjList.add(new ArrayList<>());
    }

    // Function to add an edge to the graph
    void addEdge(int u, int v) {
        adjList.get(u).add(v);
        adjList.get(v).add(u);
    }

    void BFS(int start) {
        Queue<Integer> unvisited = new LinkedList<>();
        boolean[] visited = new boolean[V];

        unvisited.add(start);
        visited[start] = true;
        while (!unvisited.isEmpty()) {
            int current = unvisited.poll();
            System.out.print(current + " ");

            for (int vertex : adjList.get(current)) {
                if (!visited[vertex]) {
                    visited[vertex] = true;
                    unvisited.add(vertex);
                }
            }

            if (unvisited.isEmpty()) {
                for (int i = 1; i < V; ++i) {
                    if (!visited[i]) {
                        System.out.println();
                        unvisited.add(i);
                        visited[i] = true;
                        break;
                    }
                }
            }
        }
    }
}

public class Assignment1 {
    public static void main(String[] args) {
        if (args.length == 0) {
            System.out.println("Usage: java Assignment1 <input_file>");
            return;
        }

        String filename = args[0];
        try {
            Scanner fileScanner = new Scanner(new File(filename));
            int graphCount = 1;
            while (fileScanner.hasNextLine()) {
                String line = fileScanner.nextLine();
                System.out.println("Graph " + graphCount++);

                Scanner lineScanner = new Scanner(line);
                int vertices = lineScanner.nextInt();

                ArrayList<Edge> edges = new ArrayList<>();
                while (lineScanner.hasNext()) {
                    String edgeString = lineScanner.next();
                    edgeString = edgeString.replaceAll("[\\(\\)]", "");
                    String[] edgeVertices = edgeString.split(",");
                    Edge edge = new Edge();
                    edge.x = Integer.parseInt(edgeVertices[0].trim());
                    edge.y = Integer.parseInt(edgeVertices[1].trim());
                    edges.add(edge);
                }

                // Create a graph
                Graph g = new Graph(vertices);

                for (Edge edge : edges) {
                    g.addEdge(edge.x, edge.y);
                }

                g.BFS(0);

                System.out.println(); // For separating outputs of different graphs
            }
        } catch (FileNotFoundException e) {
            System.err.println("File not found: " + filename);
        }
    }
}

